sensorserviceutils/sensorserviceutils.go

```diff
@@ -333,20 +467,228 @@ func (p parser) extractActiveConnInfo() error {
+		uid, err := strconv.Atoi(result["uid"])
+		if err != nil {
+			p.parsingErrs = append(p.parsingErrs,
+				fmt.Errorf("%s: error parsing uid %v for line %v: %v",
+					parseRegErrStr, result["uid"], l, err))
+			continue
+		}
+		identifier := fmt.Sprintf("%d,%s", sensorNumber, packageName)
+
+		var value string
+		if isAdd {
+			// Currently processing an activated statement.
+			samplingPeriod, err := strconv.Atoi(result["samplingPeriod"])
+			if err != nil {
+				p.parsingErrs = append(p.parsingErrs, fmt.Errorf(
+					"%s: error parsing samplingPeriod %v for line %v: %v",
+					parseRegErrStr, result["samplingPeriod"], l, err))
+				continue
+			}
+			batchingPeriod, err := strconv.Atoi(result["batchingPeriod"])
+			if err != nil {
+				p.parsingErrs = append(p.parsingErrs, fmt.Errorf(
+					"%s: error parsing batchingPeriod %v for line %v: %v",
+					parseRegErrStr, result["batchingPeriod"], l, err))
+				continue
+			}
+			value = fmt.Sprintf("%d,%d,%s,%d,%d", sensorNumber,
+				uid, packageName, samplingPeriod, batchingPeriod)
+
+			_, exist := p.history[identifier]
+			if !exist {
+				// If there is no history of de-activating a subscription,
+				// the subscription has to be active.
+				if conn, isActiveConn := p.activeConns[identifier]; isActiveConn {
+					// For active connection, set current time as the end time
+					// for the ongoing subscription event.
+					p.csvState.Print(sensorRegisDesc, "string", timestamp,
+						referenceTimestamp, value, "")
+					conn.HasRecord = true
+					p.activeConns[identifier] = conn
+				} else {
+					p.sensorErrs = append(p.sensorErrs, fmt.Errorf(
+						"[Invalid Activation]: connection between pkg(%s) and sensor(%d) should be active",
+						packageName, sensorNumber))
+				}
+			} else {
+				// A de-activation statement for the subscription event is seen.
+				eventInfo := p.history[identifier]
+				if eventInfo.StartMs != -1 {
+					// A previous de-activation statement for this connection
+					// has paired up with an activation statement. The current
+					// activation statement is an extra one.
+					p.sensorErrs = append(p.sensorErrs, fmt.Errorf(
+						"[Multiple Activation]: for pkg(%s) and sensor(%d)",
+						packageName, sensorNumber))
+				} else {
+					// The current activation statement can pair up with a
+					// previous de-activation statement to complete a
+					// subscription event.
+					eventInfo.StartMs = timestamp
+					eventInfo.SamplingPeriod = int32(samplingPeriod)
+					eventInfo.BatchingPeriod = int32(batchingPeriod)
+					p.csvState.Print(sensorRegisDesc, "string",
+						timestamp, eventInfo.EndMs, value, "")
+				}
+			}
+		} else {
+			// Currently processing an de-activation statement.
+			if event, exist := p.history[identifier]; exist {
+				if event.StartMs == -1 {
+					p.sensorErrs = append(p.sensorErrs, fmt.Errorf(
+						"[Multiple De-Activation]: for pkg(%s) and sensor(%d)",
+						packageName, sensorNumber))
```

Comment:
What if the activation was dropped off from the history.

Reply:
So for both **Multiple De-activation** and **Multiple Activation** errors, we want to consider whether the corresponding activate/de-activate action was dropped off from the history.

On top of my head, there are several cases that will result in the missing actions in history:

1. We get an error when parsing the action statement.
It is less likely that we are only having one parsing error for one action statement in the registration history. So I did not handle this case.

2. The sensorservice dump section did not capture the action so no record was found.
It is possible to check if the sensorservice dump history matches the system log history for sensors. It is less likely that we are missing the action in both the sensorservice dump and the system log history and also in other layers. I am acknowledged of this solution and I will consider this once I incorporate the system log into the historian tool.
In this case where the action was only missing in one layer but not others, we will still have a valid subscription event, but an error should be generated to acknowledge the developer about the mismatch.

3. The registration history section has some ordering issue.
_Note that the previous registration history records the subscription event in reverse chronological order._
Consider this following example:

> 17:08:09 - 0x01 pid=2 uid=3 package=A
> 17:01:20 - 0x01 pid=2 uid=3 package=A
> 17:07:09 + 0x01 pid=2 uid=3 package=A samplingPeroid=1000us batchingPeriod=0us

Because the events are not ordered, we will get a "Multiple De-activation" error. One solution I can think of is to store all individual events based on timestamp while parsing the history section. Then we process and analyze all actions and detect errors.
One assumption I have for the input data is that all events are strictly following the reverse chronological order, so I am not handling errors raised by this case for now.

4. The parser misses a line when reading in the history and the action was dropped off from history.
This case is closely related to the robustness of my sensor. I haven't considered it before. On top of my head, I don't have a solution for now.

Reply:
The events should be ordered. This is a bug if they are not!!

I am worried about the most common case where a deactivation is present and activation is not because it was a while ago and no longer present in history.

Reply:
In the case where you only see a deactivation statement but not an activation statement, I am currently neglecting this subscription event and not showing it on the visualizer since we cannot get any information simply based on the sensorservice dump. However, once we have the system log information, it is possible to create a complete subscription event.

Another thing I can do is: I can still collect such subscription event and visualize it, but with an additional tag notifying that we cannot determine whether an error exists for this event since we are missing the activation info.

Reply:
I'd prefer that it be visualized. 2 special cases

1. No activation present + deactivation present -- This should be visualized from the oldest time in registration history to deactivation time.
2. No activation or deactivation time but active connection present -- This should be visualized as active from the oldest time in the bugreport to the current time

Reply:
I agree with what you said about the first case. 
For the second case: I think it makes more sense if we visualize the subscription also as "active from**the oldest time in registration history**". If we have a gigantic bugreport, it makes less sense to say that the subscription is active this whole time. 

I will make sure to handle this while I am working on the visualizer.